src/apps/containers.clj

```diff
@@ -585,34 +599,210 @@
+  [port-uuid]
+  (first (select container-ports (where {:id (uuidify port-uuid)}))))
+
+(defn port?
+  "Returns true if the port exists."
+  [port-uuid]
+  (pos? (count (select container-ports (where {:id (uuidify port-uuid)})))))
+
+(defn port-mapping?
+  "Returns true if the the specific combination of fields exists in the
+   database."
+  [settings-uuid host-port container-port bind-to-host]
+  (pos? (count (select container-ports
+                      (where (and (= :container_settings_id (uuidify settings-uuid))
+                                  (= :host_port host-port)
+                                  (= :container_port container-port)
+                                  (= :bind_to_host bind-to-host)))))))
+
+(defn port-mapping
+  [settings-uuid host-port container-port bind-to-host]
+  (first (select container-ports
+                 (where (and (= :container_settings_id (uuidify settings-uuid))
+                             (= :host_port host-port)
+                             (= :container_port container-port)
+                             (= :bind_to_host bind-to-host))))))
+
+(defn settings-has-port?
+  [settings-uuid port-uuid]
+  (pos? (count (select container-ports
+                       (where {:container_settings_id (uuidify settings-uuid)
+                               :id                    (uuidify port-uuid)})))))
+
+(defn add-port
+  [settings-uuid port-map]
+  (let [host-port      (:host_port port-map)
+        container-port (:container_port port-map)
+        bind-to-host   (:bind_to_host port-map)]
+    (if (port-mapping? settings-uuid host-port container-port bind-to-host)
+      (throw (Exception. (str "port mapping already exists: " settings-uuid " " host-port " " container-port " " bind-to-host))))
+    (insert container-ports
+            (values (merge
+                     (select-keys port-map [:host_port :container_port :bind_to_host])
+                     {:container_settings_id (uuidify settings-uuid)})))))
+
+(defn modify-port
+  [settings-uuid port-uuid port-map]
+  (if-not (port? port-uuid)
+    (throw (Exception. (str "port does not exist: " port-uuid))))
+  (sql/update container-ports
+    (set-fields (merge {:container_settings_id (uuidify settings-uuid)}
+                       (select-keys port-map [:host_port :container_port :bind_to_host])))
+    (where {:id (uuidify port-uuid)})))
+
+(defn delete-port
+  [port-uuid]
+  (when (port? port-uuid)
+    (delete container-ports (where {:id (uuidify port-uuid)}))))
+
+(defn tool-port
+  "Returns a map containing port information."
+  [tool-uuid port-uuid]
+  (when (tool-has-settings? tool-uuid)
+    (let [settings-uuid (tool-settings-uuid tool-uuid)]
+      (when (settings-has-port? settings-uuid port-uuid)
+        (dissoc (port port-uuid) :container_settings_id)))))
+
+(defn add-tool-port
+  [tool-uuid port-map]
+  (when-not (tool-has-settings? tool-uuid)
+    (throw (Exception. (str "Tool " tool-uuid " does not have a container."))))
```

Comment:
I think that either `cxu/not-found` or `cxu/bad-request` could be used here, depending on the type of message you want to return.